File: android/src/androidTest/java/org/ligi/passandroid/ThePastLocationsStore.java
package org.ligi.passandroid;

import android.content.Context;
import android.content.SharedPreferences;
import android.support.test.rule.ActivityTestRule;
import javax.inject.Inject;
import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.ligi.passandroid.model.PastLocationsStore;
import org.ligi.passandroid.ui.PassViewActivity;
import static android.support.test.InstrumentationRegistry.getInstrumentation;
import static org.assertj.core.api.Assertions.assertThat;

public class ThePastLocationsStore {

    @Rule
    public ActivityTestRule<PassViewActivity> rule = new ActivityTestRule<>(PassViewActivity.class);

    @Inject
    Tracker tracker;

    private SharedPreferences sharedPreferences;

    @Before
    public void setUp() {
        App.setComponent(DaggerTestComponent.builder().testModule(new TestModule()).build());
        sharedPreferences = getInstrumentation().getContext().getSharedPreferences("" + (System.currentTimeMillis() / 100000), Context.MODE_PRIVATE);
        TestApp.component().inject(this);
    }

    @After
    public void tearDown() {
        sharedPreferences.edit().clear().apply();
    }

    @Test
    public void testPastLocationsStoreShouldNeverContainMoreThanMaxElements() {
        PastLocationsStore tested = new PastLocationsStore(sharedPreferences, tracker);

        for (int i = 0; i < PastLocationsStore.MAX_ELEMENTS * 2; i++) {
            tested.putLocation("" + i);
        }

        assertThat(tested.getLocations().size()).isEqualTo(PastLocationsStore.MAX_ELEMENTS);

    }

    @Test
    public void testPastLocationsStoreShouldStoreOnlyOneOfAKind() {
        PastLocationsStore tested = new PastLocationsStore(sharedPreferences, tracker);

        for (int i = 0; i < 3; i++) {
            tested.putLocation("foo");
        }

        assertThat(tested.getLocations()).containsOnly("foo");

    }

}

File: android/src/androidTest/java/org/ligi/passandroid/ThePastLocationsStore.kt
package org.ligi.passandroid

import android.content.Context
import android.content.SharedPreferences
import androidx.core.content.edit
import androidx.test.platform.app.InstrumentationRegistry
import org.assertj.core.api.Assertions.assertThat
import org.junit.After
import org.junit.Rule
import org.junit.Test
import org.ligi.passandroid.model.PastLocationsStore
import org.ligi.passandroid.ui.PassViewActivity
import org.ligi.trulesk.TruleskActivityRule
import org.mockito.Mock
import org.mockito.MockitoAnnotations

class ThePastLocationsStore {

    @get:Rule
    var rule = TruleskActivityRule(PassViewActivity::class.java) {
        TestApp.populatePassStoreWithSinglePass()
        MockitoAnnotations.initMocks(this)
    }

    @Mock
    lateinit var tracker: Tracker

    private val prefs: SharedPreferences by lazy { InstrumentationRegistry.getInstrumentation().context.getSharedPreferences("" + System.currentTimeMillis() / 100000, Context.MODE_PRIVATE) }

    @After
    fun tearDown() {
        prefs.edit { clear() }
    }

    @Test
    fun testPastLocationsStoreShouldNeverContainMoreThanMaxElements() {
        val tested = PastLocationsStore(prefs, tracker)

        for (i in 0 until PastLocationsStore.MAX_ELEMENTS * 2) {
            tested.putLocation("" + i)
        }

        assertThat(tested.locations.size).isEqualTo(PastLocationsStore.MAX_ELEMENTS)

    }

    @Test
    fun testPastLocationsStoreShouldStoreOnlyOneOfAKind() {
        val tested = PastLocationsStore(prefs, tracker)

        for (i in 0..2) {
            tested.putLocation("foo")
        }

        assertThat(tested.locations).containsOnly("foo")

    }

}
